morebits.js

```diff
@@ -4566,215 +4577,1725 @@ Morebits.wiki.page = function(pageName, status) {
+	 * users, or users with specific template or optout links.  Some
+	 * options are customizable, but implies `setCreateOption('recreate')`
+	 * and `setFollowRedirect(true, false)`; other options are available
+	 * via `setPageobjectFunctions`.  Makes use of:
+	 * - {@link Morebits.wiki.user#setMessage|setMessage}
+	 * - {@link Morebits.wiki.user#setSectionTitle|setSectionTitle}
+	 * - {@link Morebits.wiki.user#setNotifyBots|setNotifyBots}
+	 * - {@link Morebits.wiki.user#setNotifyIndef|setNotifyIndef}
+	 * - {@link Morebits.wiki.user#setNotifySelf|setNotifySelf}
+	 * - {@link Morebits.wiki.user#setNotifySkips|setNotifySkips}
+	 * - {@link Morebits.wiki.user#setPageobjectFunctions|setPageobjectFunctions}
+	 *
+	 * The actual processing is handled in `fnProcessNotify`.
+	 *
+	 * @param {Function} [onSuccess] - Callback function to run on success.
+	 * @param {Function} [onFailure] - Callback function to run on failure.
+	 */
+	this.notify = function(onSuccess, onFailure) {
+		ctx.onNotifySuccess = onSuccess;
+		ctx.onNotifyFailure = onFailure || emptyFunction;
+
+		if (ctx.isIPRange) {
+			ctx.statusElement.error('Cannot notify IP ranges');
+			ctx.onNotifyFailure(this);
+			return;
+		}
+		// Check underscores
+		if (ctx.notifySelf && ctx.userName === mw.config.get('wgUserName')) {
+			ctx.statusElement.error('Skipping self notification');
+			ctx.onNotifyFailure(this);
+			return;
+		}
+
+		// Ensure user is loaded
+		if (fnDontNeedLoad('notify')) {
+			fnProcessNotify.call(this);
+		} else {
+			this.load(fnProcessNotify, ctx.onNotifyFailure);
+		}
+	};
+
+	// Send the notification
+	var fnProcessNotify = function() {
+		// Empty reason, message, and token handled by Morebits.wiki.page
+		if (!ctx.exists) {
+			ctx.statusElement.error('Cannot notify the user because the user does not exist');
+			ctx.onNotifyFailure(this);
+			return;
+		}
+
+		if (ctx.notifySkipTemplates && ctx.notifySkipTemplates.length && ctx.talkTemplates && ctx.talkTemplates.length) {
+			// More efficient to do a for loop, but this is prettier?
+			var tlDups = Morebits.array.dups(ctx.talkTemplates.concat(ctx.notifySkipTemplates));
+			if (tlDups.length) {
+				ctx.statusElement.error('User talk page transcludes ' + tlDups[0] + ', aborting notification');
+				ctx.onNotifyFailure(this);
+				return;
+			}
+		} else if (ctx.notifySkipLink && ctx.talkLinks && ctx.talkLinks.length) {
+			// Should be without leading protocol; relying on mw.Uri could help
+			var elDups = Morebits.array.dups(ctx.talkLinks.concat(ctx.notifySkipLink));
+			if (elDups.length) {
+				ctx.statusElement.error('User has opted out of this notification, aborting');
+				ctx.onNotifyFailure(this);
+				return;
+			}
+
+		}
+
+		if (!ctx.notifyBots && this.isBot()) {
+			ctx.statusElement.error('User is a bot, aborting notification');
+			ctx.onNotifyFailure(this);
+			return;
+		}
+		// Clients may find this most useful iff notalk or the block isn't brand new
```

Comment:
>block isn't brand new

Would be a good feature to add, IMO; maybe only for blocks older than two weeks or something.